app/screens/TransactionAuthorization.tsx

```diff
@@ -54,38 +54,32 @@ export function TransactionAuthorization (): JSX.Element | null {
   const [status, emitEvent] = useState<Status>('INIT')
   const [attemptsRemaining, setAttemptsRemaining] = useState<number>(MAX_PASSCODE_ATTEMPT)
   const [pin, setPin] = useState<string>('')
-  const [isSubmitted, setIsSubmitted] = useState<boolean>(false)
+  const [isRetry, setIsRetry] = useState(false)
 
   // generic callbacks
   const onPinInput = useCallback((pin: string): void => {
-    setPin(pin)
     if (pin.length === PIN_LENGTH && PASSPHRASE_PROMISE_PROXY !== undefined) {
       const resolve = PASSPHRASE_PROMISE_PROXY.resolve
-      emitEvent('SIGNING')
-      setIsSubmitted(true)
       setTimeout(() => {
         resolve(pin)
         // remove proxied promised, allow next prompt() call
         PASSPHRASE_PROMISE_PROXY = undefined
       }, 50)
+      emitEvent('SIGNING')
     }
+    setPin(pin)
   }, [PASSPHRASE_PROMISE_PROXY, PASSPHRASE_PROMISE_PROXY?.resolve])
 
   const onCancel = useCallback((): void => {
-    setPin('')
-    emitEvent('CANCELLED')
-    setIsSubmitted(false)
     if (PASSPHRASE_PROMISE_PROXY !== undefined) {
-      const reject = PASSPHRASE_PROMISE_PROXY.reject
-      setTimeout(() => {
-        reject(new Error(USER_CANCELED))
-        // remove proxied promised, allow next prompt() call
-        PASSPHRASE_PROMISE_PROXY = undefined
-      }, 50)
+      PASSPHRASE_PROMISE_PROXY.reject(new Error(USER_CANCELED))
```

Comment:
do not require extra setState here, rejecting the pin promise with USER_CANCELED error lead the signing/authenticate result handler.